lib/sprite_animation.dart

```diff
@@ -4,6 +4,71 @@ import 'dart:ui';
 import 'extensions/vector2.dart';
 import 'sprite.dart';
 
+class SpriteAnimationFrameData {
+  /// Coordinates of the sprite of this Frame
+  final Vector2 srcPosition;
+
+  /// Size of the sprite of this Frame
+  final Vector2 srcSize;
+
+  /// The duration to display it, in seconds.
+  double stepTime;
+
+  SpriteAnimationFrameData({
+    @required this.srcPosition,
+    @required this.srcSize,
+    @required this.stepTime,
+  });
+}
+
+class SpriteAnimationData {
+  List<SpriteAnimationFrameData> frames;
+  final bool loop;
+
+  SpriteAnimationData(this.frames, {this.loop = true});
+
+  SpriteAnimationData.variable({
+    @required int amount,
+    @required List<double> stepTimes,
+    int amountPerRow,
+    Vector2 texturePosition,
+    Vector2 textureSize,
+    this.loop = true,
+  })  : assert(amountPerRow == null || amount >= amountPerRow),
+        assert(stepTimes != null) {
+    amountPerRow ??= amount;
+    texturePosition ??= Vector2.zero();
+    frames = List<SpriteAnimationFrameData>(amount);
+    for (int i = 0; i < amount; i++) {
+      final position = Vector2(
+        texturePosition.x + (i % amountPerRow) * textureSize.x,
+        texturePosition.y + (i ~/ amountPerRow) * textureSize.y,
+      );
+      frames[i] = SpriteAnimationFrameData(
+        stepTime: stepTimes[i],
+        srcPosition: position,
+        srcSize: textureSize,
+      );
+    }
+  }
+
+  SpriteAnimationData.sequenced({
+    @required int amount,
+    @required double stepTime,
+    int amountPerRow,
+    Vector2 texturePosition,
+    Vector2 textureSize,
+    bool loop = true,
+  }) : this.variable(
+          amount: amount,
+          amountPerRow: amountPerRow,
+          texturePosition: texturePosition,
+          textureSize: textureSize,
+          loop: loop,
+          stepTimes: stepTime != null ? List.filled(amount, stepTime) : null,
```

Comment:
stepTimes is required non-null on variable. so this doesn't make a lot of sense. Probably should be list.filled(amount, stepTime ?? _DEFAULT_STEP_TIME)

Reply:
I explicitly added null here because the other constructor validates this on the assert. It is a pitty that bridge constructors on dart can't have asserts as well.

I think I prefer this solution over a default stepTime, what do you think?

Reply:
so my understand of this code is that if stepTime here is null it will pass null to the other constructor and throw a NPE. but I haven't deeply analyzed the code, if this works it's fine for me :)